pkg/network/ebpf/c/protocols/classification/routing.h

```diff
@@ -0,0 +1,98 @@
+#ifndef __PROTOCOL_ROUTING_H
+#define __PROTOCOL_ROUTING_H
+
+#include "ktypes.h"
+#include "protocols/classification/defs.h"
+#include "protocols/classification/stack-helpers.h"
+
+// These macros are meant to improve the readability of the `__get_next_program` function below
+// LAYER_ENTRYPOINT(program) designates the first (socket-filter) program to be executed for a given layer
+#define LAYER_ENTRYPOINT(program)                               \
+    do {                                                        \
+        if (current_program == CLASSIFICATION_PROG_UNKNOWN) {   \
+            return program;                                     \
+        }                                                       \
+    } while(0)
+
+// LAYER_TRANSITION designates a sequence of socket-filter programs within a given layer
+#define LAYER_TRANSITION(from, to)              \
+    do {                                        \
+        if (current_program == from) {          \
+            return to;                          \
+        }                                       \
+    } while(0)
+
+#define LAYER_END(layer_mask)                           \
+    do {                                                \
+        current_program = CLASSIFICATION_PROG_UNKNOWN;  \
+        *known_layers |= layer_mask;                    \
+    } while(0)
+
+// This function essentially encodes all routing aspects of tail-calls. For
+// example, if this function gets called from `CLASSIFICATION_QUEUES_PROG` the
+// return value will be likely `CLASSIFICATION_DBS_PROG` (which is the next
+// program that belongs to the same stack layer), but that depends whether or
+// not the application layer protocol is known at the time of the call. When a
+// certain protocol layer is known, the function "skips" to the entry-point of
+// the next layer and so forth.
+static __always_inline classification_prog_t __get_next_program(usm_context_t *usm_ctx) {
+    u16 *known_layers = &usm_ctx->routing_known_layers;
+    classification_prog_t current_program = usm_ctx->routing_current_program;
+
+    if (*known_layers&LAYER_APPLICATION_BIT) {
+        goto api;
+    }
+
+    // Add Application-layer routing here
+    LAYER_ENTRYPOINT(CLASSIFICATION_QUEUES_PROG);
+    LAYER_TRANSITION(CLASSIFICATION_QUEUES_PROG, CLASSIFICATION_DBS_PROG);
```

Comment:
wouldn't it be better to perform enum arithmetics instead of explicitly using the enum values?

Reply:
I'll try to play with the idea 👍 Would love to simplify this code further